### PR TITLE
Refactor Call Graph Producer

### PR DESCRIPTION
**Pull request by gdrosos**
## Description
This PR aims to refactor the Python Call Graph Producer in the following ways:

* Removed the Call graph converter phase since now PyCG produces Call Graphs in the correct Fasten format.
* Applied the separation of concerns design pattern in cg-producer by separating all the kafka related functionalities from the call-graph generation processes.
* Published the Call graph Generation class to PyPI, in order to be used as a standalone library by our partners.
* Added github action workflow for publishing  the Call Graph Producer docker image to Github Registry under the Fasten namespace

## Motivation 
The call graph producer contained all the functionalities (e.g. kafka processes, call graph generation processes) in a single file, making it hard to comprehend. 
It also contained a call graph format conversion mechanism which was redundant after the merge of [this pull request](https://github.com/vitsalis/PyCG/pull/27).
 Finally, up until now, the process of downloading, unzipping, and generating the call graph of a pypi coordinate was a utility only feasible in a kafka production environment, meaning that there was no  standalone tool enabling the automatic process of a a signle PyPI coordinate.

## Testing
Tested the changes both locally, in docker and also by using the [pycg-producer library](https://pypi.org/project/pycg-producer/0.0.5/) through [pip](https://pypi.org/project/pycg-producer/0.0.5/)

### CODE REVIEW

File: .github/workflows/publish-docker-cg-producer.yaml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+          IMAGE_ID=docker.pkg.github.com/${{ github.repository }}/$IMAGE_NAME
+          # Change all uppercase to lowercase
+          IMAGE_ID=$(echo $IMAGE_ID | tr '[A-Z]' '[a-z]')
+          
+          # Strip git ref prefix from version
+          VERSION=$(echo "${{ github.ref }}" | sed -e 's,.*/\(.*\),\1,')
+          # Strip "v" prefix from tag name
+          [[ "${{ github.ref }}" == "refs/tags/"* ]] && VERSION=$(echo $VERSION | sed -e 's/^v//')
+          # Use Docker `latest` tag convention
+          [ "$VERSION" == "master" ] && VERSION=${GITHUB_SHA::8}
```
Comment (vitsalis) [inline .github/workflows/publish-docker-cg-producer.yaml]: Wasn't `master` renamed to `main`?

File: .gitignore [hunk 2]
```diff
@@ -1 +1,161 @@
 *.swp
+# Byte-compiled / optimized / DLL files
+__pycache__/
```
Comment (vitsalis) [inline .gitignore:3]: Why do we need such an extensive list of things to ignore? Most of this stuff is not related to our repository.

File: cg-producer/Dockerfile [hunk 3]
```diff
@@ -39,6 +39,8 @@ RUN mkdir /cggen
 # move files
 COPY ./PyCG/ /pycg
 COPY ./entrypoint.py /cggen/entrypoint.py
+COPY ./pycg_producer /cggen/pycg_producer
```
Comment (vitsalis) [inline cg-producer/Dockerfile:42]: Please use dashes (`-`) as separators on directory names instead of low dashes (`_`).

File: cg-producer/setup.cfg [hunk 4]
```diff
@@ -0,0 +1,2 @@
+[metadata]
+description-file = pypi-readme.md
```
Comment (vitsalis) [inline cg-producer/setup.cfg]: Is it a common convention for this particular file name to not be capitalized? If not, I would suggest renaming it to `PyPI-README.md`.

File: cg-producer/pypi-readme.md [hunk 5]
```diff
@@ -0,0 +1,23 @@
+Library for generating PyPI Call Graphs through PyCG
```
Comment (vitsalis) [inline cg-producer/pypi-readme.md]: Please use `#` to describe section headers. Makes it more readable. (e.g. `#` is Header1, `##` is Header2 etc.)

File: cg-producer/pypi-readme.md [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+====================================================
+
+This library can be used to produce the call graph of a package distributed through pip along with saving its
+source code and the call graph in json format in a specified directory.
```
Comment (vitsalis) [inline cg-producer/pypi-readme.md]: `JSON`

File: cg-producer/pypi-readme.md [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+Usage
+-----
+
+Example usage after installing the specific package through ```pip```:
```
Comment (vitsalis) [inline cg-producer/pypi-readme.md]: For inline code listings a single ` is enough.

File: cg-producer/pypi-readme.md [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+``` python
```
Comment (vitsalis) [inline cg-producer/pypi-readme.md]: Extra space after ```

File: cg-producer/pypi-readme.md [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+from pycg_producer.producer import CallGraphGenerator
+
+coord = { "product": "pycg-stitch",
+          "version": "0.0.8",
+          "version_timestamp": "2000",
+          "requires_dist": []}
+generator = CallGraphGenerator("directoryName", coord)
+print(generator.generate())
+```
+The CallGraphGenerator class recieves as input the name of the directory where the source code and the call graph json will be stored, along with a python dictionary containing the product name and version of the specific pypi coordinate.
```
Comment (vitsalis) [inline cg-producer/pypi-readme.md]: `JSON`

`Python`

`PyPI`

File: cg-producer/setup.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
+#!/usr/bin/env python
+
+from distutils.core import setup
+
+def get_long_desc():
+    with open("pypi-readme.md", "r") as readme:
```
Comment (vitsalis) [inline cg-producer/setup.py]: Why not just name this `README.md`?

File: cg-producer/entrypoint.py [hunk 11]
```diff
@@ -522,9 +65,13 @@ def consume(self):
                 release
             ))
 
-            generator = CallGraphGenerator(self.out_topic, self.err_topic,
-                    self.source_dir, self.producer, release)
-            generator.generate()
+            generator = CallGraphGenerator(self.source_dir, release)
+            output = generator.generate()
+
+            if output["Status"]=="Success":
```
Comment (vitsalis) [inline cg-producer/entrypoint.py]: You can use constants instead of strings. Makes the code more maintainable, e.g.:

```
SUCCESS = "Success"
...
if output["Status"] == SUCCESS:
...
```

File: cg-producer/entrypoint.py [hunk 12]
```diff
@@ -522,9 +65,13 @@ def consume(self):
...
+                self.producer.send(self.out_topic, json.dumps(output["Output"]))
+            elif output["Status"]=="Fail":
```
Comment (vitsalis) [inline cg-producer/entrypoint.py]: Same here.

File: cg-producer/Dockerfile (hunk 3 above)
Comment (gdrosos) [reply cg-producer/Dockerfile:42]: Since this directory hosts a Python package, it should follow the [Pep8](https://pep8.org/#package-and-module-names) naming conversions. The package names should have short, all-lowercase names. The underscope is accepted but is [discouraged](https://pep8.org/#package-and-module-names). So, we can rename the directory maybe to `pycgproducer` if you think that it would suit better.

File: .github/workflows/publish-docker-cg-producer.yaml (hunk 1 above)
Comment (gdrosos) [reply .github/workflows/publish-docker-cg-producer.yaml]: Fixed on [551ae39](https://github.com/fasten-project/pypi-tools/pull/6/commits/551ae3998a000f848c97704948e916a36e41e556)

File: .gitignore (hunk 2 above)
Comment (gdrosos) [reply .gitignore:3]: Fixed on [f332710](https://github.com/fasten-project/pypi-tools/pull/6/commits/f33271055d61e1a13d02a7f9729732f11aa62083), I kept only some files that were generated when building and publishing the library to pypi.

File: cg-producer/pypi-readme.md (hunk 9 above)
Comment (gdrosos) [reply cg-producer/pypi-readme.md]: Fixed on [1efab6e](https://github.com/fasten-project/pypi-tools/pull/6/commits/1efab6ebab16447fa84ad504621bfecf3e283821)

File: cg-producer/pypi-readme.md (hunk 8 above)
Comment (gdrosos) [reply cg-producer/pypi-readme.md]: Fixed on [1efab6e](https://github.com/fasten-project/pypi-tools/pull/6/commits/1efab6ebab16447fa84ad504621bfecf3e283821)

File: cg-producer/entrypoint.py (hunk 11 above)
Comment (gdrosos) [reply cg-producer/entrypoint.py]: You are right, changed on [8bbdc5f](https://github.com/fasten-project/pypi-tools/pull/6/commits/8bbdc5f69710d4942fb0a4257dee02b97caaa797).

File: cg-producer/entrypoint.py (hunk 12 above)
Comment (gdrosos) [reply cg-producer/entrypoint.py]: Fixed on [8bbdc5f](https://github.com/fasten-project/pypi-tools/pull/6/commits/8bbdc5f69710d4942fb0a4257dee02b97caaa797)

File: cg-producer/setup.py (hunk 10 above)
Comment (gdrosos) [reply cg-producer/setup.py]: The reason behind this additional file is that we have a main [Readme file](https://github.com/fasten-project/pypi-tools/blob/refactor-pycg-producer/cg-producer/README.md) which  includes instructions on running the Kafka Call Graph Generator. But we wanted to create an additional description file for the  pycg-producer library, which is distributed through PyPI and needed a description.

**Comment by gdrosos** [comment]
Hello @vitsalis, thanks for the immediate response and for all the insightful comments!
I tried to address all the comments. 
Please let me know if anything needs to be fixed or improved!

File: cg-producer/Dockerfile (hunk 3 above)
Comment (vitsalis) [reply cg-producer/Dockerfile:42]: Ah, I see. In that case it's better to keep the underscore for readability.

**Comment by vitsalis** [comment]
LGTM

File: cg-producer/pypi-readme.md (hunk 5 above)
Comment (gdrosos) [reply cg-producer/pypi-readme.md]: Fixed on [1efab6e](https://github.com/fasten-project/pypi-tools/pull/6/commits/1efab6ebab16447fa84ad504621bfecf3e283821)

File: cg-producer/pypi-readme.md (hunk 6 above)
Comment (gdrosos) [reply cg-producer/pypi-readme.md]: Fixed on [01f7959](https://github.com/fasten-project/pypi-tools/pull/6/commits/01f79591f34c0281a9104a8f12e0475b5f9a5234)

File: cg-producer/pypi-readme.md (hunk 7 above)
Comment (gdrosos) [reply cg-producer/pypi-readme.md]: Fixed on [01f7959](https://github.com/fasten-project/pypi-tools/pull/6/commits/01f79591f34c0281a9104a8f12e0475b5f9a5234)

File: cg-producer/setup.cfg (hunk 4 above)
Comment (gdrosos) [reply cg-producer/setup.cfg]: No, there is no such convention, changed the file name accordingly on [12c573e](https://github.com/fasten-project/pypi-tools/pull/6/commits/12c573e8ab02b3848bd09ca765d8f02a335a32c5)

**Comment by gdrosos** [comment]
Thanks! I am merging.
